### ISSUE TITLE
Cookies missing on very first request if website has service worker with fetch event handler

### ISSUE THREAD

**Issue by steffenweber** [uBlockOrigin/uMatrix-issues#155]
### Prerequisites

- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
    - [x] I also searched the existing issues at <https://github.com/gorhill/uMatrix/issues>
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uMatrix](https://old.reddit.com/r/uMatrix/)
- I tried to reproduce the issue when...
    - [x] uMatrix is the only extension
    - [x] uMatrix with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uMatrix
- [x] I checked the [documentation](https://github.com/gorhill/uMatrix/wiki) to understand that the issue I report is not a normal behavior
- [x] I used the logger to rule out that the issue is caused by my ruleset

### Description

If a website has a service worker with a non-trivial "fetch" event handler (i.e. it doesn't just do nothing but uses the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) to return a response) then the very first request to a website is sent without any cookies. (For more details on service workers in general see the MDN docs [Using Service Workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers).)

I've tested with uMatrix 1.3.16 and 1.3.17rc0 on Firefox 66.0.5, Firefox 68.0b2 and Firefox Nightly.

### A specific URL where the issue occurs

https://www.computerbase.de/

### Steps to Reproduce

1. Open https://www.computerbase.de/ to initialize the service worker and to get a "service-worker-init-date" cookie (any cookie would do)
2. In the same tab, open any other website, for example https://www.google.com/
3. In the same tab, open https://www.computerbase.de/ again

What I'd expect to happen: The GET request for https://www.computerbase.de/ should have the "service-worker-init-date" cookie set.

What actually happens: The GET request for https://www.computerbase.de/ is sent without any cookies (as can be seen in the "Network" tab of the Firefox developer tools).

Notice that if you now click on any local link (or reload the page) then the cookies _are_ sent. It's just the _very first_ request that is sent without cookies.

The issue goes away if I either uninstall/disable uMatrix or if I delete the "fetch" event handler from the computerbase.de service worker (I'm the admin of computerbase.de). It seems like the very first request is somehow treated in a special way and gets stripped of all cookies by uMatrix.

This issue affects the login cookies of computerbase.de and was reported to me in a private conversation by one of our users (nickname "Whistl0r") and then discussed in a little more detail here: https://www.computerbase.de/forum/threads/login-probleme-seit-14-mai-der-aktive-benutzer-hat-sich-geaendert.1872153/page-2

### Ruleset

I have tested with a fresh Firefox profile and have not changed the default configuration of Firefox or uMatrix in any way. These are the contents of the "My rules" tab in the uMatrix settings:

```
https-strict: behind-the-scene false
matrix-off: about-scheme true
matrix-off: behind-the-scene true
matrix-off: chrome-extension-scheme true
matrix-off: chrome-scheme true
matrix-off: moz-extension-scheme true
matrix-off: opera-scheme true
matrix-off: vivaldi-scheme true
matrix-off: wyciwyg-scheme true
noscript-spoof: * true
referrer-spoof: * true
referrer-spoof: behind-the-scene false
* * * block
* * css allow
* * frame block
* * image allow
* 1st-party * allow
* 1st-party frame allow
```

### Supporting evidence

Screenshot showing that there is no "Cookie" header when following the "Steps to Reproduce":
![Screenshot from 2019-05-20 15-15-05-fs8](https://user-images.githubusercontent.com/1138129/58024615-e3069d00-7b12-11e9-995b-9b27aa22fc0e.png)

Screenshot showing that there is a "Cookie" header if you reload the page or click on any local link:
![Screenshot from 2019-05-20 15-45-30-fs8](https://user-images.githubusercontent.com/1138129/58026391-a9d02c00-7b16-11e9-97bd-a98c01ee7c53.png)

### Your environment

* uMatrix version: 1.3.16 / 1.3.17rc0
* Browser Name and version: Firefox 66.0.5 / 68.0b2 / Nightly
* Operating System and version: Gentoo Linux (x86_64)

**Comment by gwarser** [comment]
Wondering if it can be this again https://github.com/uBlockOrigin/uMatrix-issues/issues/72 -> https://bugzilla.mozilla.org/show_bug.cgi?id=1499523

**Comment by gwarser** [comment]
`doc` request missing:
![image](https://user-images.githubusercontent.com/886325/58032614-41d41280-7b23-11e9-978f-c4364a7c2799.png)

**Comment by gorhill** [comment]
No `COOKIE` header removal is reported in the logger -- so uMatrix is not removing these cookies.

**Comment by gwarser** [comment]
It works when cookies are allowed in global scope.

**Comment by gorhill** [comment]
Never mind, I can reproduce now, I always forget that the Firefox setting _"Delete cookies and site data when Firefox is closed"_ prevent service workers from being created.

**Comment by gorhill** [comment]
So the fix is to lookup and use `documentUrl` for context purpose when available rather than the `behind-the-scene` context. uMatrix is way behind uBO with all the filtering context improvements.

**Comment by gwarser** [comment]
Did you noticed something is still wrong with cookie logging?

- sw is set, cookies blocked in global scope
- open google.com in new tab -> only google cookies in µM logger
- navigate to https://www.computerbase.de/ - no new cookie entries
- navigate back to google.com -> google cookies + computerbase.de cookie as 3p to google.

![image](https://user-images.githubusercontent.com/886325/58042074-fe38d300-7b39-11e9-97db-2a113985a678.png)

**Comment by steffenweber** [comment]
I can confirm that this issue is gone in uMatrix 1.3.17rc1. Thank you very much for the extremely quick fix! :+1:

**Comment by steffenweber** [comment]
@gorhill This issue was fixed in uMatrix 1.3.17rc1 / 1.3.17rc2 but has reappeared in uMatrix 1.3.17rc3. Could you please have another look?

**Comment by gorhill** [comment]
The real issue is the Firefox bug pointed out by @gwarser above: <https://bugzilla.mozilla.org/show_bug.cgi?id=1499523>.

My first fix was incorrect, I overlooked the fact I need to use the root context, and `documentUrl` may not be that of the root context, hence this led to the regression reported in <https://github.com/uBlockOrigin/uMatrix-issues/issues/159>, and fixing that regression brought back the issue here.

Since this is a really Firefox issue, I have to investigate whether it can at least be mitigated without introducing new regressions.

**Comment by steffenweber** [comment]
FYI: The related Ghostery issue: https://github.com/ghostery/ghostery-extension/issues/382

It contains a description of their attempt to workaround this Firefox bug. Maybe that's interesting for you.

**Comment by uBlock-user** [comment]
@steffenweber A new fix landed in 1.3.17rc4, try that and see if you can still reproduce.

**Comment by steffenweber** [comment]
1.3.17rc4 fixes the issue again. Thank you! :+1:
